launchlib/memory.go

```diff
@@ -54,18 +63,58 @@ func (c CGroupMemoryLimit) MemoryLimitInBytes() (uint64, error) {
 		return 0, errors.Wrap(err, "failed to get memory cgroup path")
 	}
 
-	memLimitFilepath := filepath.Join(memoryCGroupPath, memLimitName)
-	memLimitFile, err := c.fs.Open(convertToFSPath(memLimitFilepath))
+	isCGroupV2, err := IsCGroupV2(c.fs)
 	if err != nil {
-		return 0, errors.Wrapf(err, "unable to open memory.limit_in_bytes at expected location: %s", memLimitFilepath)
+		return 0, errors.Wrap(err, "failed to determine cgroup version")
 	}
-	memLimitBytes, err := io.ReadAll(memLimitFile)
-	if err != nil {
-		return 0, errors.Wrapf(err, "unable to read memory.limit_in_bytes")
+
+	var memLimitBytes []byte
+	if isCGroupV2 {
+		memLimitBytes, err = c.readCGroupV2MemoryLimit(memoryCGroupPath)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to get cgroupv2 memory limit")
+		}
+	} else {
+		memLimitBytes, err = c.readCGroupV1MemoryLimit(memoryCGroupPath)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to get cgroupv1 memory limit")
+		}
+	}
+
+	memLimitStr := strings.TrimSpace(string(memLimitBytes))
+	if memLimitStr == "max" {
+		return 0, errors.New("memory limit is set to max")
 	}
-	memLimit, err := strconv.Atoi(strings.TrimSpace(string(memLimitBytes)))
+
+	memLimit, err := strconv.Atoi(memLimitStr)
 	if err != nil {
-		return 0, errors.New("unable to convert memory.limit_in_bytes value to expected type")
+		return 0, errors.New("unable to convert memory limit value to expected type")
 	}
 	return uint64(memLimit), nil
 }
+
+func (c CGroupMemoryLimit) readCGroupV1MemoryLimit(memoryCGroupPath string) ([]byte, error) {
+	cgroupV1MemLimitFilepath := filepath.Join(memoryCGroupPath, cgroupV1MemLimitName)
+	cgroupV1MemLimitFile, err := c.fs.Open(convertToFSPath(cgroupV1MemLimitFilepath))
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to open memory.limit_in_bytes at expected location: %s", cgroupV1MemLimitFile)
+	}
+	memLimitBytes, err := io.ReadAll(cgroupV1MemLimitFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to read memory.limit_in_bytes")
+	}
+	return memLimitBytes, nil
+}
+
+func (c CGroupMemoryLimit) readCGroupV2MemoryLimit(memoryCGroupPath string) ([]byte, error) {
+	cgroupV2MemLimitFilepath := filepath.Join(memoryCGroupPath, cgroupV2MemLimitName)
+	cgroupV2MemLimitFile, err := c.fs.Open(convertToFSPath(cgroupV2MemLimitFilepath))
```

Comment:
From [these docs](https://pkg.go.dev/io/fs#FS), I think we need to call `cgroupV2MemLimitFile.close()` at some point?